### ISSUE TITLE
ENH: C-API function for slicing ndarray[i:j]

### ISSUE THREAD

**Issue by jbrockmendel** [numpy/numpy#23934]
### Proposed new feature or change:

I discussed this with @seberg a few weeks ago.  In pandas we have a hotspot in the cython code where we slice an ndarray.  AFAICT the slicing itself is fast, but having to make a python-space call is slow.  bc the relevant call happens inside a loop, this adds up.

Specific request: a numpy C-API function for ndarray[i:j]

**Comment by ianharris** [comment]
I came across this issue and thought implementing a simple function to perform a slice operation using the NumPy C-API might be a nice way to learn about creating NumPy extensions. With @seberg's suggestion that a small C function might be of use in mind, I created: https://github.com/ianharris/c-ndarray-slice

Hopefully it is something that is somewhat close to what was in mind in the discussion above, and is also reasonably robust and useful.

**Comment by jbrockmendel** [comment]
thanks for taknig a look ian.  IIUC the implementation you posted will make a copy, is that correct?  For a slice I expect a view.

**Comment by rkern** [comment]
Given all of the ways that this needs to be restricted from general slicing/indexing in order to get something fast, and the potential benefit that a single object that continually gets modified in the loop instead might provide instead, I'm going to suggest that this is not a candidate for a new API function. It seems like something that's good to implement near where it's needed and those extra restrictions make sense.

**Comment by ianharris** [comment]
@jbrockmendel yes, it is a copy. I naively hadn't thought about a view. I'll need to take another look to see how a view could be done.

For your use case I'm guessing you can slice along any dimension? And you mention processing in a loop, do you take slices of slices? Or could you assume that an array is only ever sliced once?

**Comment by jbrockmendel** [comment]
> For your use case I'm guessing you can slice along any dimension? And you mention processing in a loop, do you take slices of slices? Or could you assume that an array is only ever sliced once?

The use case I have in mind is always 2D and slicing on axis=0.  And the loop isn't on a single ndarray, but over a sequence of ndarrays.

**Comment by ianharris** [comment]
@jbrockmendel the zero axis simplifies things. Had to do a bit of research to understand some NumPy internals but I think I have a plan now. I'll write some updated code for this tomorrow.

**Comment by ianharris** [comment]
An updated version that no longer requires a copy: https://github.com/ianharris/c-ndarray-slice/tree/view-0-axis

**Comment by jbrockmendel** [comment]
Thanks @ianharris.

@WillAyd if this doesn't land in numpy, what do you think of vendoring it?  Unfortunately we just [refactored away](https://github.com/pandas-dev/pandas/pull/54568) the main place where we would have used it, so making use of it would require re-implementing specialized Block subclasses.  Then we'd implement NumpyBlock._slice that would use this directly.  The main place where I think it might make a difference would be in _get_block_for_concat_plan where we call slice_block_columns in a (nested) loop.  Unfortunately I don't have a good estimate on how big the perf impact would be.

**Comment by WillAyd** [comment]
I don't see a problem doing this in pandas itself. We already have the capsule extension module so this (and other similar functions) could be added there to be exposed to our existing extensions
